rtc_tenclave/src/dh/sessions.rs

```diff
@@ -0,0 +1,328 @@
+    TInit: RtcDhInitiator,
+{
+    sessions: RwLock<HashMap<u64, Arc<AtomicSession<TResp>>>>,
+    _phantom_init: PhantomData<TInit>,
+}
+
+impl<TResp, TInit> DhSessions<TResp, TInit>
+where
+    TResp: RtcDhResponder,
+    TInit: RtcDhInitiator,
+{
+    fn get(&self, id: &u64) -> Option<Arc<AtomicSession<TResp>>> {
+        self.sessions
+            .read()
+            .expect("RwLock poisoned")
+            .get(id)
+            .map(Clone::clone)
+    }
+
+    fn lock_write(&self) -> RwLockWriteGuard<HashMap<u64, Arc<AtomicSession<TResp>>>> {
+        self.sessions.write().expect("RwLock poisoned")
+    }
+
+    pub fn get_active(&self, id: &u64) -> Option<Arc<Mutex<ProtectedChannel>>> {
+        match self.get(id)?.as_ref() {
+            AtomicSession::Active(x) => Some(x.clone()),
+            _ => None,
+        }
+    }
+
+    fn take_in_progress(&self, id: &u64) -> Option<TResp> {
+        let mut sessions = self.lock_write();
+
+        if matches!(sessions.get(id)?.as_ref(), AtomicSession::InProgress(_)) {
+            match Arc::try_unwrap(sessions.remove(id)?) {
+                Ok(AtomicSession::InProgress(resp)) => Some(resp),
+                Ok(_) => unreachable!(),
+                Err(_) => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn close_session(&self, id: &u64) -> () {
+        let mut sessions = self.lock_write();
+        sessions.insert(*id, Arc::new(AtomicSession::Closed));
+        ()
+    }
+
+    /// Creates and sets an active session between this enclave and the enclave with `id` using
+    /// the `key`.
+    ///
+    /// # Valid Operations (It is the responsibility of the caller to ensure these hold)
+    /// None -> Active = Ok
+    /// Closed -> Active = Ok
+    /// InProgress -> Active = Err if it is the same session
+    ///   - In progress value needs to be removed from the map before using it to finalize a session
+    /// Active -> Active = Err
+    ///   - A session should be closed and then recreated to prevent resetting the nonce counter
+    ///     with the same key.
+    fn set_active(
+        &self,
+        id: u64,
+        key: Secret<AlignedKey>,
+    ) -> Result<Arc<Mutex<ProtectedChannel>>, sgx_status_t> {
+        let channel = Arc::new(Mutex::new(ProtectedChannel::init(key)));
+        self.lock_write()
+            .insert(id, Arc::new(AtomicSession::Active(channel.clone())));
+        Ok(channel)
+    }
+
+    /// Sets an in_progress session for a responding enclave linked to the provided enclave.
+    ///
+    /// # Valid Operations (It is the responsibility of the caller to ensure these hold)
+    /// InProgress -> InProgress = Ok
+    /// Closed -> InProgress = Ok
+    /// Active -> In Progress = Ok if keying material differs
+    fn set_in_progress(&self, id: &u64, responder: TResp) -> Result<(), sgx_status_t> {
+        let _result = self
+            .lock_write()
+            .insert(*id, Arc::new(AtomicSession::InProgress(responder)));
```

Comment:
This `id` probably makes more sense as a copy?

```suggestion
    fn set_in_progress(&self, id: u64, responder: TResp) -> Result<(), sgx_status_t> {
        let _result = self
            .lock_write()
            .insert(id, Arc::new(AtomicSession::InProgress(responder)));
```